### ISSUE TITLE
Memory issue when doing select

### ISSUE THREAD

**Issue by lhoestq** [huggingface/nlp#395]
As noticed in #389, the following code loads the entire wikipedia in memory.

```python
import nlp
w = nlp.load_dataset("wikipedia", "20200501.en", split="train")
w.select([0])
```

This is caused by [this line](https://github.com/huggingface/nlp/blob/master/src/nlp/arrow_dataset.py#L626) for some reason, that tries to serialize the function with all the wikipedia data with it.

It's not the case with `.map` or `.filter`.
However functions that are based on `.select` like `.shuffle`, `.shard`, `.train_test_split`, `.sort` are affected.

**Comment by narwhal3301** [comment]
If you think you have it -- https://narwhal3301.github.io/
